include/seastar/util/log-impl.hh

```diff
@@ -50,40 +50,43 @@ private:
     void realloc_buffer();
 
 public:
+    // inserter_iterator is designed like std::back_insert_iterator:
+    // operator*, operator++ and operator++(int) are no-ops,
+    // and all the work happens in operator=, which pushes a character
+    // to the buffer.
+    // The iterator stores no state of its own.
+    //
+    // inserter_iterator is supposed to be used as an output_iterator.
+    // That is, assignment is expected to alternate with incrementing.
     class inserter_iterator {
     public:
         using iterator_category = std::output_iterator_tag;
         using difference_type = std::ptrdiff_t;
-        using value_type = char;
-        using pointer = char*;
-        using reference = char&;
+        using value_type = void;
+        using pointer = void;
+        using reference = void;
 
     private:
         log_buf* _buf;
-        char* _current;
 
     public:
-        explicit inserter_iterator(log_buf& buf) noexcept : _buf(&buf), _current(_buf->_current) { }
-        inserter_iterator(const inserter_iterator& o) noexcept : _buf(o._buf), _current(o._current) { }
+        explicit inserter_iterator(log_buf& buf) noexcept : _buf(&buf) { }
 
-        reference operator*() {
-            if (__builtin_expect(_current == _buf->_end, false)) {
+        inserter_iterator& operator=(char c) {
```

Comment:
Here I don't understand why operator= doesn't just return void (which, incidentally, is your "reference" type!)
Sure, if we return void, then
i=j=3
won't work, but it's not supposed to work... we only want to support *(i++)=3 et al, not i=j=3, so the latter code might as well just fail, no?

Reply:
Returning void would be fine too. I didn't think too hard about this, I simply copied the design of `std::back_insert_iterator`.